app/controllers/url_redirects_controller.rb

```diff
@@ -87,6 +87,15 @@ def download_product_files
     else
       # Non-JSON requests to this controller route pass an array with a single product file ID for `product_file_ids`
       @product_file = product_files.first
+
+      if @product_file.must_be_pdf_stamped? && !@url_redirect.alive_stamped_pdfs.exists?(product_file_id: @product_file.id)
+        StampPdfForPurchaseJob.perform_async(@url_redirect.purchase_id)
+
+        flash[:notice] = "The PDF will be emailed to you shortly!"
+        redirect_to url_redirect_download_page_path(@url_redirect.token)
+        return
+      end
```

Comment:
starts the asynchronous stamping operation and sends an alert that pdf will be emailed shortly